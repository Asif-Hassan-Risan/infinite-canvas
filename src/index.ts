import {sayHello} from "./say-hello"

sayHello("emile");