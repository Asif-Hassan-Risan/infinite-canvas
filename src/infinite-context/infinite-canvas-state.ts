import { ViewBox } from "../viewbox";

export class InfiniteCanvasState implements CanvasState{
	constructor(private viewBox: ViewBox){}
	public restore():void{
		this.viewBox.restoreState();
	}
	public save(): void{
		this.viewBox.saveState();
	}
}