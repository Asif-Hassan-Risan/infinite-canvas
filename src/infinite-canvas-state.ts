import { Transformation } from "./transformation";
import { DrawingInstruction } from "./drawing-instruction";

export class InfiniteCanvasState implements DrawingInstruction{
    constructor(
        public readonly lineWidth: number,
        public readonly lineDashOffset: number,
        public readonly lineDash: number[],
        public readonly fillStyle: string | CanvasGradient | CanvasPattern,
        public readonly strokeStyle: string | CanvasGradient | CanvasPattern
    ){

    }
    public apply(context: CanvasRenderingContext2D, transformation: Transformation): void{
        context.lineWidth = this.lineWidth * transformation.scale;
        context.lineDashOffset = this.lineDashOffset * transformation.scale;
        context.setLineDash(this.lineDash.map(s => s * transformation.scale));
        context.fillStyle = this.fillStyle;
        context.strokeStyle = this.strokeStyle;
    }
    public setLineWidth(lineWidth: number): InfiniteCanvasState{
        return new InfiniteCanvasState(
            lineWidth,
            this.lineDashOffset,
            this.lineDash,
            this.fillStyle,
            this.strokeStyle
        );
    }
    public setLineDashOffset(lineDashOffset: number): InfiniteCanvasState{
        return new InfiniteCanvasState(
            this.lineWidth,
            lineDashOffset,
            this.lineDash,
            this.fillStyle,
            this.strokeStyle
        );
    }
    public setLineDash(lineDash: number[]): InfiniteCanvasState{
        return new InfiniteCanvasState(
            this.lineWidth,
            this.lineDashOffset,
            lineDash,
            this.fillStyle,
            this.strokeStyle
        );
    }
    public setFillStyle(fillStyle: string | CanvasGradient | CanvasPattern): InfiniteCanvasState{
        return new InfiniteCanvasState(
            this.lineWidth,
            this.lineDashOffset,
            this.lineDash,
            fillStyle,
            this.strokeStyle
        );
    }
    public setStrokeStyle(strokeStyle: string | CanvasGradient | CanvasPattern): InfiniteCanvasState{
        return new InfiniteCanvasState(
            this.lineWidth,
            this.lineDashOffset,
            this.lineDash,
            this.fillStyle,
            strokeStyle
        );
    }
    public static default(): InfiniteCanvasState{
        return new InfiniteCanvasState(
            1,
            0,
            [],
            "#000",
            "#000"
        );
    }
}