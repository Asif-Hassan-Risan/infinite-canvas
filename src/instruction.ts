import { Transformation } from "./transformation";

export interface Instruction{
    apply(context: CanvasRenderingContext2D, transformation: Transformation): void;
}