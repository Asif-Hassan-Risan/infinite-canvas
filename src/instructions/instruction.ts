import { Transformation } from "../transformation";

export declare type Instruction = (context: CanvasRenderingContext2D, transformation: Transformation) => void;