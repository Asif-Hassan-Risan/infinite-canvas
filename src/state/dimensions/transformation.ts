import { Transformation } from "../../transformation";
import { InfiniteCanvasStateInstanceDimension } from "./infinite-canvas-state-instance-dimension";
import { Instruction } from "../../instructions/instruction";
import { TypedStateInstanceDimension } from "./typed-state-instance-dimension";

class TransformationDimension extends InfiniteCanvasStateInstanceDimension<"transformation">{
    protected valuesAreEqual(oldValue: Transformation, newValue: Transformation): boolean{
        return oldValue.equals(newValue);
    }
    protected changeToNewValue(newValue: Transformation): Instruction{
        return (context: CanvasRenderingContext2D, transformation: Transformation) => {
            const {a, b, c, d, e, f} = transformation.inverse().before(newValue).before(transformation);
            context.setTransform(a, b, c, d, e, f);
        }
    }
}
export const transformation: TypedStateInstanceDimension<Transformation> = new TransformationDimension("transformation");