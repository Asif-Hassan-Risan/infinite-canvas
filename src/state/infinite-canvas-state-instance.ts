import { StateChange } from "./state-change";
import { Transformation } from "../transformation";
import { Instruction } from "../instructions/instruction";
import { ChainableStateChange } from "./chainable-state-change";

export class InfiniteCanvasStateInstance {
    constructor(
        public fillStyle: string | CanvasGradient | CanvasPattern,
        public lineWidth: number,
        public lineDash: number[],
        public strokeStyle: string | CanvasGradient | CanvasPattern,
        public lineDashOffset: number,
        public transformation: Transformation
    ){}
    private noStateChange(): StateChange<InfiniteCanvasStateInstance>{
        return {
            newState: this,
            instruction: undefined
        };
    }
    private static lineDashesAreEqual(one: number[], other: number[]): boolean{
        if(one.length !== other.length){
            return false;
        }
        for(let i=0; i<one.length;i++){
            if(one[i] !== other[i]){
                return false;
            }
        }
        return true;
    }

    public convertToState(other: InfiniteCanvasStateInstance): StateChange<InfiniteCanvasStateInstance>{
        return new ChainableStateChange<InfiniteCanvasStateInstance>(this, [])
            .change(s => s.setFillStyle(other.fillStyle))
            .change(s => s.setLineWidth(other.lineWidth))
            .change(s => s.setLineDash(other.lineDash))
            .change(s => s.setStrokeStyle(other.strokeStyle))
            .change(s => s.setLineDashOffset(other.lineDashOffset))
            .change(s => s.setTransformation(other.transformation));
    }

    public equals(other: InfiniteCanvasStateInstance): boolean{
        return this.fillStyle === other.fillStyle &&
               this.lineWidth === other.lineWidth &&
               InfiniteCanvasStateInstance.lineDashesAreEqual(this.lineDash, other.lineDash) &&
               this.strokeStyle === other.strokeStyle &&
               this.lineDashOffset === other.lineDashOffset &&
               this.transformation.equals(other.transformation);
    }

    public setFillStyle(value: string | CanvasGradient | CanvasPattern): StateChange<InfiniteCanvasStateInstance>{
        if(value === this.fillStyle){
            return this.noStateChange();
        }
        return {
            newState: new InfiniteCanvasStateInstance(value, this.lineWidth, this.lineDash, this.strokeStyle, this.lineDashOffset, this.transformation),
            instruction: (context: CanvasRenderingContext2D, transformation: Transformation) => {
                context.fillStyle = value;
            }
        };
    }

    public setLineWidth(lineWidth: number): StateChange<InfiniteCanvasStateInstance>{
        if(lineWidth === this.lineWidth){
            return this.noStateChange();
        }
        return {
            newState: new InfiniteCanvasStateInstance(this.fillStyle, lineWidth, this.lineDash, this.strokeStyle, this.lineDashOffset, this.transformation),
            instruction: (context: CanvasRenderingContext2D, transformation: Transformation) => {
                context.lineWidth = transformation.scale * lineWidth;
            }
        }
    }

    public setLineDash(lineDash: number[]): StateChange<InfiniteCanvasStateInstance>{
        if(InfiniteCanvasStateInstance.lineDashesAreEqual(this.lineDash, lineDash)){
            return this.noStateChange();
        }
        return {
            newState: new InfiniteCanvasStateInstance(this.fillStyle, this.lineWidth, lineDash, this.strokeStyle, this.lineDashOffset, this.transformation),
            instruction: (context: CanvasRenderingContext2D, transformation: Transformation) => {
                context.setLineDash(lineDash.map(d => d * transformation.scale));
            }
        };
    }

    public setStrokeStyle(strokeStyle: string | CanvasGradient | CanvasPattern): StateChange<InfiniteCanvasStateInstance>{
        if(this.strokeStyle === strokeStyle){
            return this.noStateChange();
        }
        return {
            newState: new InfiniteCanvasStateInstance(this.fillStyle, this.lineWidth, this.lineDash, strokeStyle, this.lineDashOffset, this.transformation),
            instruction: (context: CanvasRenderingContext2D, transformation: Transformation) => {
                context.strokeStyle = strokeStyle;
            }
        };
    }

    public setLineDashOffset(lineDashOffset: number): StateChange<InfiniteCanvasStateInstance>{
        if(this.lineDashOffset === lineDashOffset){
            return this.noStateChange();
        }
        return {
            newState: new InfiniteCanvasStateInstance(this.fillStyle, this.lineWidth, this.lineDash, this.strokeStyle, lineDashOffset, this.transformation),
            instruction: (context: CanvasRenderingContext2D, transformation: Transformation) => {
                context.lineDashOffset = lineDashOffset * transformation.scale;
            }
        }
    }

    public addTransformation(transformation: Transformation): StateChange<InfiniteCanvasStateInstance>{
        if(transformation.equals(Transformation.identity)){
            return this.noStateChange();
        }
        const newTransformation: Transformation = transformation.before(this.transformation);
        return {
            newState: new InfiniteCanvasStateInstance(this.fillStyle, this.lineWidth, this.lineDash, this.strokeStyle, this.lineDashOffset, newTransformation),
            instruction: (context: CanvasRenderingContext2D, transformation: Transformation) => {
                const {a, b, c, d, e, f} = transformation.inverse().before(newTransformation).before(transformation);
                context.setTransform(a, b, c, d, e, f);
            }
        };
    }

    public setTransformation(givenTransformation: Transformation): StateChange<InfiniteCanvasStateInstance>{
        if(this.transformation.equals(givenTransformation)){
            return this.noStateChange();
        }
        return {
            newState: new InfiniteCanvasStateInstance(this.fillStyle, this.lineWidth, this.lineDash, this.strokeStyle, this.lineDashOffset, givenTransformation),
            instruction: (context: CanvasRenderingContext2D, transformation: Transformation) => {
                const {a, b, c, d, e, f} = transformation.inverse().before(givenTransformation).before(transformation);
                context.setTransform(a, b, c, d, e, f);
            }
        };
    }
    public static default: InfiniteCanvasStateInstance = new InfiniteCanvasStateInstance('#000', 1, [], "#000", 0, Transformation.identity);
    public static setDefault: Instruction = (context: CanvasRenderingContext2D, transformation: Transformation) => {
        context.lineWidth = transformation.scale;
    };
}