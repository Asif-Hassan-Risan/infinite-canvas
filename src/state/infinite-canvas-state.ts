import { InfiniteCanvasStateInstance } from "./infinite-canvas-state-instance";
import { StateChange } from "./state-change";
import { ChainableStateChange } from "./chainable-state-change";

export class InfiniteCanvasState{
    constructor(public current: InfiniteCanvasStateInstance, public stack: InfiniteCanvasStateInstance[] = []){}
    public static default: InfiniteCanvasState = new InfiniteCanvasState(InfiniteCanvasStateInstance.default, []);
    public withChangedState(change: (state: InfiniteCanvasStateInstance) => StateChange<InfiniteCanvasStateInstance>): StateChange<InfiniteCanvasState>{
        const stateChange: StateChange<InfiniteCanvasStateInstance> = change(this.current);
        return {
            newState: new InfiniteCanvasState(stateChange.newState, this.stack),
            instruction: stateChange.instruction
        };
    }
    public save(): StateChange<InfiniteCanvasState>{
        return {
            newState: new InfiniteCanvasState(this.current, (this.stack || []).concat([this.current])),
            instruction: (context: CanvasRenderingContext2D) => {context.save();}
        };
    }
    public restore(): StateChange<InfiniteCanvasState>{
        if(!this.stack || this.stack.length === 0){
            return {
                newState: this,
                instruction: undefined
            };
        }
        const topOfStack: InfiniteCanvasStateInstance = this.stack[this.stack.length - 1];
        return {
            newState: new InfiniteCanvasState(topOfStack, this.stack.slice(0, this.stack.length - 1)),
            instruction: (context: CanvasRenderingContext2D) => {context.restore();}
        };
    }
    public convertTo(other: InfiniteCanvasState): StateChange<InfiniteCanvasState>{
        let change: ChainableStateChange<InfiniteCanvasState> = new ChainableStateChange(this, []);
        const indexOfHighestCommon: number = InfiniteCanvasState.findIndexOfHighestCommon(this.stack, other.stack);
        let currentIndex: number = indexOfHighestCommon + 1;
        for(let i: number = this.stack.length - 1; i > indexOfHighestCommon; i--){
            change = change.change(s => s.restore());
        }
        while(currentIndex < other.stack.length){
            change = change.change(s => s.withChangedState(ss => ss.convertToState(other.stack[currentIndex])));
            change = change.change(s => s.save());
            currentIndex++;
        }
        change = change.change(s => s.withChangedState(ss => ss.convertToState(other.current)));
        return change;
    }

    private static findIndexOfHighestCommon(one: InfiniteCanvasStateInstance[], other: InfiniteCanvasStateInstance[]): number{
        let result: number = 0;
        let lowestIndex: number = Math.min(one.length - 1, other.length - 1);
        while(result <= lowestIndex){
            if(!one[result].equals(other[result])){
                return result - 1;
            }
            result++;
        }
        return result - 1;
    }
}