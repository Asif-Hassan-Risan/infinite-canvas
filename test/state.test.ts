import { InfiniteCanvasStateInstance } from "../src/state/infinite-canvas-state-instance";
import { StateChange } from "../src/state/state-change";
import { logInstruction } from "./log-instruction";
import { Instruction } from "../src/instructions/instruction";
import { Transformation } from "../src/transformation";
import { InfiniteCanvasState } from "../src/state/infinite-canvas-state";

describe("a default state", () => {
    let state: InfiniteCanvasState;

    beforeEach(() => {
        state = InfiniteCanvasState.default;
    });

    describe("that changes and saves", () => {

        beforeEach(() => {
            state = state.withChangedState(s => s.setFillStyle("#f00")).newState.save().newState;
        });

        describe("and then changes in two different ways", () => {
            let oneWay: InfiniteCanvasState;
            let otherWay: InfiniteCanvasState;

            beforeEach(() => {
                oneWay = state.withChangedState(s => s.setStrokeStyle("#f00")).newState;
                otherWay = state.withChangedState(s => s.setFillStyle("#00f")).newState;
            });

            describe("and then one way is converted to the other way", () => {
                let change: StateChange<InfiniteCanvasState>;

                beforeEach(() => {
                    change = oneWay.convertTo(otherWay);
                });

                it("should result in instructions that change the state, but no restoring", () => {
                    expect(change.instruction).toBeTruthy();
                    expect(logInstruction(change.instruction)).toMatchSnapshot();
                });
            });
        });

        describe("and is then converted to another that also changed and saved", () => {
            let other: InfiniteCanvasState;
            let change: StateChange<InfiniteCanvasState>;

            beforeEach(() => {
                other = InfiniteCanvasState.default.withChangedState(s => s.setFillStyle("#00f")).newState.save().newState;
                change = state.convertTo(other);
            });

            it("should have created intructions to restore, change and save", () => {
                expect(change.instruction).toBeTruthy();
                expect(logInstruction(change.instruction)).toMatchSnapshot();
            });
        });

        describe("and is then converted to another that changed, saved and then changed again", () => {
            let other: InfiniteCanvasState;
            let change: StateChange<InfiniteCanvasState>;

            beforeEach(() => {
                other = InfiniteCanvasState.default
                    .withChangedState(s => s.setFillStyle("#00f")).newState
                    .save().newState
                    .withChangedState(s => s.setFillStyle("#ff0")).newState;
                change = state.convertTo(other);
            });

            it("should have created intructions to restore, change, save and change again", () => {
                expect(change.instruction).toBeTruthy();
                expect(logInstruction(change.instruction)).toMatchSnapshot();
            });
        });
    });

    describe("that is changed to a non-default state that differs on some fields but not all", () => {
        let change: StateChange<InfiniteCanvasState>;
        let otherState: InfiniteCanvasState;

        beforeEach(() => {
            otherState = new InfiniteCanvasState(new InfiniteCanvasStateInstance(
                "#000",                                 //same
                2,                                      //different
                [1,2],                                  //different
                "#f00",                                 //different
                0,                                      //same
                new Transformation(2, 0, 0, 2, 0, 0)    //different
            ));
            change = state.convertTo(otherState);
        });

        it("should have created an instruction that sets the fields that differ", () => {
            const createdInstruction: Instruction = change.instruction;
            expect(createdInstruction).toBeTruthy();
            expect(logInstruction(createdInstruction)).toMatchSnapshot();
        });

        it("should have created a change with a new state that is equal to the state that it changed to", () => {
            const newState: InfiniteCanvasStateInstance = change.newState.current;
            expect(newState.fillStyle).toBe(otherState.current.fillStyle);
            expect(newState.lineWidth).toBe(otherState.current.lineWidth);
            expect(newState.lineDash).toEqual(otherState.current.lineDash);
            expect(newState.strokeStyle).toBe(otherState.current.strokeStyle);
            expect(newState.lineDashOffset).toBe(otherState.current.lineDashOffset);
            expect(newState.transformation.equals(otherState.current.transformation)).toBe(true);
        });

        describe("and is then changed to that same state again", () => {
            let newChange: StateChange<InfiniteCanvasState>;

            beforeEach(() => {
                newChange = change.newState.convertTo(otherState);
            });

            it("should not have created an instruction", () => {
                expect(newChange.instruction).toBeFalsy();
            });
        });
    });

    describe("that sets the fill style", () => {
        let change: StateChange<InfiniteCanvasState>;

        beforeEach(() => {
            change = state.withChangedState(s => s.setFillStyle("#f00"));
        });

        it("should have created an instruction", () => {
            const createdInstruction: Instruction = change.instruction;
            expect(createdInstruction).toBeTruthy();
            expect(logInstruction(createdInstruction)).toMatchSnapshot();
        });

        describe("and then sets it again on the new state", () => {
            let newChange: StateChange<InfiniteCanvasState>;

            beforeEach(() => {
                newChange = change.newState.withChangedState(s => s.setFillStyle("#f00"));
            });

            it("should not have created an instruction", () => {
                const createdInstruction: Instruction = newChange.instruction;
                expect(createdInstruction).toBeFalsy();
            });
        });
    });
});