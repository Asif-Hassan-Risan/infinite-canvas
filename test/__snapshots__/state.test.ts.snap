// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`a default state that changes and saves and is then converted to another that also changed and saved should have created intructions to restore, change and save 1`] = `
Array [
  "context.restore()",
  "context.fillStyle = \\"#00f\\"",
  "context.save()",
]
`;

exports[`a default state that changes and saves and is then converted to another that changed, saved and then changed again should have created intructions to restore, change, save and change again 1`] = `
Array [
  "context.restore()",
  "context.fillStyle = \\"#00f\\"",
  "context.save()",
  "context.fillStyle = \\"#ff0\\"",
]
`;

exports[`a default state that changes and saves and then changes in two different ways and then one way is converted to the other way should result in instructions that change the state, but no restoring 1`] = `
Array [
  "context.fillStyle = \\"#00f\\"",
  "context.strokeStyle = \\"#000\\"",
]
`;

exports[`a default state that is changed to a non-default state that differs on some fields but not all should have created an instruction that sets the fields that differ 1`] = `
Array [
  "context.lineWidth = 2",
  "context.setLineDash([1,2])",
  "context.strokeStyle = \\"#f00\\"",
  "context.setTransform(2,0,0,2,0,0)",
]
`;

exports[`a default state that sets the fill style should have created an instruction 1`] = `
Array [
  "context.fillStyle = \\"#f00\\"",
]
`;
